dex/src/main/scala/com/wavesplatform/dex/db/OrderDB.scala

```diff
@@ -1,86 +1,101 @@
-  def getOrderInfo(id: Order.Id): Option[FinalOrderInfo]
-  def transactionsByOrder(orderId: ByteStr): Seq[ExchangeTransaction]
+trait OrderDb[F[_]] {
+  def containsInfo(id: Order.Id): F[Boolean]
+  def status(id: Order.Id): F[OrderStatus.Final]
+  def saveOrderInfo(id: Order.Id, sender: Address, oi: OrderInfo[OrderStatus.Final]): F[Unit]
+  def saveOrder(o: Order): F[Unit]
+  def get(id: Order.Id): F[Option[Order]]
+  def getFinalizedOrders(owner: Address, maybePair: Option[AssetPair]): F[Seq[(Order.Id, OrderInfo[OrderStatus])]]
+  def getOrderInfo(id: Order.Id): F[Option[FinalOrderInfo]]
+  def transactionsByOrder(orderId: ByteStr): F[Seq[ExchangeTransaction]]
 }
 
-object OrderDB {
+object OrderDb {
   case class Settings(maxOrders: Int)
 
-  def apply(settings: Settings, db: DB): OrderDB = new OrderDB with ScorexLogging {
-    override def containsInfo(id: Order.Id): Boolean = db.readOnly(_.has(DbKeys.orderInfo(id)))
+  def levelDb[F[_]](settings: Settings, levelDb: LevelDb[F])(implicit F: Monad[F]): OrderDb[F] = new OrderDb[F] {
 
-    override def status(id: Order.Id): OrderStatus.Final = db.readOnly { ro =>
+    override def containsInfo(id: Order.Id): F[Boolean] =
+      levelDb.readOnly(_.has(DbKeys.orderInfo(id)))
+
+    override def status(id: Order.Id): F[OrderStatus.Final] = levelDb.readOnly { ro =>
       ro.get(DbKeys.orderInfo(id)).fold[OrderStatus.Final](OrderStatus.NotFound)(_.status)
     }
 
-    override def saveOrder(o: Order): Unit = db.readWrite { rw =>
+    override def saveOrder(o: Order): F[Unit] = levelDb.readWrite { rw =>
       val k = DbKeys.order(o.id())
       if (!rw.has(k))
         rw.put(k, Some(o))
     }
 
-    override def get(id: Order.Id): Option[Order] = db.readOnly(_.get(DbKeys.order(id)))
+    override def get(id: Order.Id): F[Option[Order]] = levelDb.readOnly(_.get(DbKeys.order(id)))
 
-    override def saveOrderInfo(id: Order.Id, sender: Address, oi: FinalOrderInfo): Unit = {
+    override def saveOrderInfo(id: Order.Id, sender: Address, oi: FinalOrderInfo): F[Unit] = {
       val orderInfoKey = DbKeys.orderInfo(id)
-      if (!db.has(orderInfoKey))
-        db.readWrite { rw =>
-          val newCommonSeqNr = rw.inc(DbKeys.finalizedCommonSeqNr(sender))
-          rw.put(DbKeys.finalizedCommon(sender, newCommonSeqNr), Some(id))
-
-          val newPairSeqNr = rw.inc(DbKeys.finalizedPairSeqNr(sender, oi.assetPair))
-          rw.put(DbKeys.finalizedPair(sender, oi.assetPair, newPairSeqNr), Some(id))
-          if (newPairSeqNr > settings.maxOrders) // Indexes start with 1, so if maxOrders=100 and newPairSeqNr=101, we delete 1 (the first)
-            rw.get(DbKeys.finalizedPair(sender, oi.assetPair, newPairSeqNr - settings.maxOrders))
-              .map(DbKeys.order)
-              .foreach(x => rw.delete(x))
-
-          rw.put(orderInfoKey, Some(oi))
-        }
+      for {
+        hasKey <- levelDb.has(orderInfoKey)
+        _ <-
+          F.whenA(!hasKey) {
```

Comment:
Maybe, define it in separate method, to make for-comprehension clearer?
Like this:

```
for {
  hasKey <- levelDb.has(orderInfoKey)
  _   <- myMultilineFunction(hasKey)
} yield ...
```

Reply:
resolved